### PR TITLE
Sign in with credential implementation

### PR DESCRIPTION
**Pull request by mksmdvdff**
I don't have enough experience with js, so review it more carefully, please.

P.S. I can't use typealiases because AuthCredential is abstract class in Android and simple class in iOS

### CODE REVIEW

**Comment by mksmdvdff** [comment]
@vanniktech which kind of test do you want? As I see, not all methods covered with tests right now. 
I can add some platform specific tests for Android and iOS with EmailAuthCredential, but it differs from current approach.

**Comment by vanniktech** [comment]
That you can sign in using email + password. There are already tests for auth.

**Comment by mksmdvdff** [comment]
@vanniktech Ok, as I understand - there are two ways to add tests for this case
1. Add dependencies on `firebase auth` for `test` source sets. It will allow to create credentials inside of them. At least - in `androidAndroidTest` and `iOSTest`
2. Add EmailCredentials to common code. It looks strange for me to add single implementation just for testing, but it's still an option
Which one do you like more?

**Comment by mksmdvdff** [comment]
Added tests. Build failed with unknown reason

**Comment by mksmdvdff** [comment]
Done

**Comment by mksmdvdff** [comment]
HI, can I get come feedback? Current branch can be built locally, but fails to build on commit. Can we do smth?

**Comment by nbransby** [comment]
@Reedyuk do you know why this build is failing?

File: firebase-auth/build.gradle.kts [hunk 1]
```diff
@@ -97,11 +97,6 @@ kotlin {
                 implementation(project(":firebase-common"))
             }
         }
-        val androidMain by getting {
-            dependencies {
-                api("com.google.firebase:firebase-auth:19.1.0")
-            }
-        }
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-auth/build.gradle.kts]: Fixed

File: firebase-auth/src/androidAndroidTest/kotlin/dev/gitlive/firebase/auth/auth.kt [hunk 2]
```diff
@@ -6,8 +6,9 @@
 package dev.gitlive.firebase.auth
 
 import androidx.test.platform.app.InstrumentationRegistry
+import com.google.firebase.auth.EmailAuthProvider
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-auth/src/androidAndroidTest/kotlin/dev/gitlive/firebase/auth/auth.kt]: Fixed

File: firebase-app/build.gradle.kts [hunk 3]
```diff
@@ -90,6 +90,11 @@ kotlin {
                 api("com.google.firebase:firebase-common:19.2.0")
             }
         }
+        val androidAndroidTest by getting {
+            dependencies {
+                api("com.google.firebase:firebase-common:19.2.0")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-app/build.gradle.kts]: Fixed

File: firebase-auth/src/jsTest/kotlin/dev/gitlive/firebase/auth/auth.kt [hunk 4]
```diff
@@ -10,3 +10,8 @@ import kotlinx.coroutines.promise
 actual val context: Any = Unit
 
 actual fun runTest(test: suspend () -> Unit) = GlobalScope.promise { test() }.unsafeCast<Unit>()
+
+actual fun createEmailCredential(
+    email: String,
+    password: String
+): AuthCredential? = null
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-auth/src/jsTest/kotlin/dev/gitlive/firebase/auth/auth.kt]: Fixed

File: firebase-auth/src/iosTest/kotlin/dev/gitlive/firebase/auth/auth.kt [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 package dev.gitlive.firebase.auth
 
+import cocoapods.FirebaseAuth.FIREmailAuthProvider
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-auth/src/iosTest/kotlin/dev/gitlive/firebase/auth/auth.kt]: Fixed

File: firebase-auth/src/commonMain/kotlin/dev/gitlive/firebase/auth/auth.kt [hunk 6]
```diff
@@ -40,12 +42,18 @@ expect class FirebaseUser {
...
+expect class FirebaseAuthInvalidCredentialsException : FirebaseAuthException
+expect class FirebaseAuthInvalidUserException : FirebaseAuthException
+expect class FirebaseAuthRecentLoginRequiredException : FirebaseAuthException
+expect class FirebaseAuthUserCollisionException : FirebaseAuthException
+expect class FirebaseAuthWebException : FirebaseAuthException
+
+expect class AuthCredential
+
+expect object EmailAuthProvider{
+    fun withEmailAndPassword(email: String, password: String): AuthCredential
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-auth/src/commonMain/kotlin/dev/gitlive/firebase/auth/auth.kt]: Fixed

File: firebase-common/src/jsMain/kotlin/dev/gitlive/firebase/externals.kt [hunk 7]
```diff
@@ -62,15 +62,18 @@ external object firebase {
...
             fun onAuthStateChanged(nextOrObserver: (user.User?) -> Unit): () -> Unit
         }
         interface AuthResult {
             val user: user.User?
         }
-    }
+        interface AuthCredential
 
+    }
+    fun createEmailCredential(email: String, password: String) : auth.AuthCredential
```
[the remark this answers is not in the document]
Comment (mksmdvdff) [reply firebase-common/src/jsMain/kotlin/dev/gitlive/firebase/externals.kt]: Looks like i fix it, but review it one more time, please. Do I need to use abstract class here?

**Comment by mksmdvdff** [comment]
Looks like there some troubles with build - it fails for another PR with the same error. Waiting for fixes

**Comment by Reedyuk** [comment]
@nbransby - so the reason the build is failing is because for each unique branch, they get their own cache, so the issue still currently remains when fetching the firebase zip file. In this case, its unable to fetch the cache, so try's to download the zip file and then fails for whatever reason.
More investigation is needed.

**Comment by nbransby** [comment]
Ok CI fixed and tests passing, thanks for the contribution @mksmdvdff !

**Pull request merged**